### PR TITLE
Logger

### PR DESCRIPTION
**Pull request by eaardal**
## Request ID

- Track action end-to-end
- Created as early as possible
  - Carried throughout the workflow
    - Function param?
    - `context`?
- Must be unique
- Should be human recognizable
- Number + Name combo?

---

## Request ID Examples

- 1468a232-5228-billowing-river
- ef68a1c1-8ce7-twilight-resonance
- d403762b-7351-still-field
- c60a8236-a49c-crimson-grass
- e89de930-0547-empty-wood
- 7803f7b7-e684-floral-snowflake
- 9d38c468-e2e9-bold-dew
- 081ee927-636f-twilight-glitter
- 34048512-c0fa-twilight-lake
- dc186e33-9495-black-voice

---

## Request ID

- Number + Name is hard to validate
  - Is not UUID
  - Is not a standard format
  - Hard to regex
  - Do we care? :D

---

## Function param

```go
func (Service) GetCustomer(requestId string, ...) {
  ...
}
```

---

## Context

> The intended use for Context.Value, from the original blog post, is request-scoped values. A request scoped value is one derived from data in the incoming request and goes away when the request is over.

---

## Context

> Obvious request scoped data could be who is making the request (user ID), how they are making it (internal or external), from where they are making it (user IP),and how important this request should be.

---

## Context

```go
// Add value to context
requestId := GenRequestId()
bgctx := context.Background()
ctx := context.WithValue(bgctx, "REQUEST_ID", requestId)
```

<br />
```go
// Get value from context
requestId := ctx.Value("REQUEST_ID").(string)
```

---

## LoggerContext

- Implements `context.Context`
- Has `ctx.GetRequestId()`
- Must have a Request ID
- Creates a new Request ID or accepts an existing one

---

## LoggerContext

```go
func (Service) GetCustomer(ctx LoggerContext, ...) {
  ...
}
```

---

## Other request params?

- Put `CustomerID` and `UserID` on `LoggerContext`?
- If so, rename `LoggerContext`?

---

## Logger what-if

- Could we aggregate downstream logs?
  - Full end-to-end logs in every service
  - Same log output in every service
  - Better traceability

---

logger

---

## Logger

- Normalizes log output
- Logs to stdout only
- Sensible log levels: `DBG`, `INF`, `ERR`
- Caches log entries
- Dump & absorb

---

## Logger details

- Caches everything regardless of log level
- Min. log level in each application decides output
  - ServiceA with log level `DBG` will print all logs from ServiceB even if log level in ServiceB is `ERR`
- Integrates closely with `LoggerContext` and `RequestID`

---

## Implications

- All outgoing requests must send `RequestID`
- All servers must require being sent a `RequestID`
- All responses must include the cache of log entries that was produced during the handling of the corresponding request
- When a GRPC call is complete, the resulting logs must be absorbed into the caller's logs.

### CODE REVIEW

File: src/buldog/logger_context.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	return l.ctx.Value(key)
+}
+
+// Tries to find TraceID in the context's values. The TraceID must:
+// 		- exist as a value on the context under the key defined by the const 'TraceIdKey'
+// 		- must be a string with length > 0
+func (l LoggerContext) GetTraceId() string {
+	traceIdRaw := l.ctx.Value(TraceIdKey)
+	if traceIdRaw == nil {
+		log.Fatalf("logger could not extract traceId from context. Value with key '%s' is nil\n", TraceIdKey)
```
Comment (jonalu) [inline src/buldog/logger_context.go]: Usikker på om `log.Fatal` er riktig for desse tre sjekkane. Kanskje heller gi ein id, men logge issue

File: src/buldog/logger_context_mock.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import "time"
+
+type LoggerContextMock struct {
+	DeadlineFn          func() (deadline time.Time, ok bool)
+	DoneFn              func() <-chan struct{}
+	ErrFn               func() error
+	ValueFn             func(key interface{}) interface{}
+	GetTraceIdFn        func() string
+	IsGetTraceIdInvoked bool
```
Comment (jonalu) [inline src/buldog/logger_context_mock.go]: Her kan du godt legge på `Is*Invoked` for alle metodane. Så slepp vi inn igjen her å legge det til når vi skal teste i konsumentappane.

File: src/buldog/logger_mock.go [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+		},
+		IsInfoDataInvoked: false,
+		IsInfoDataInvokedWithMsg: func(string) (bool, string) {
+			return false, ""
+		},
+		ErrorFn: func(ILoggerContext, string, ...interface{}) {
+
+		},
+		IsErrorInvoked: false,
+		IsErrorInvokedWithMsg: func(string) (bool, string) {
```
Comment (jonalu) [inline src/buldog/logger_mock.go]: Desse `Is*InvokedWithMsg` returnerer `(bool, string)` Ikkje tydelig kva denne param 2 er. Er det faktisk invoked msg? Kanskje vi her kan bruke named attrs

File: src/buldog/logger_mock.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	m.IsErrorInvokedWithMsg = func(t string) (bool, string) {
+		actual := fmt.Sprintf(text, args...)
+		return t == actual, actual
+	}
+	m.ErrorFn(ctx, text, args...)
+}
+
+func (m *LoggerMock) ErrorData(ctx ILoggerContext, data interface{}, text string, args ...interface{}) {
+	m.IsErrorDataInvoked = true
+	m.IsErrorDataInvokedWithMsg = func(t string) (bool, string) {
```
Comment (jonalu) [inline src/buldog/logger_mock.go]: Det var litt snedig gjort. Slepp å mellomlagre `InvokedMsg` 👏

File: src/buldog/trace_id.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package buldog
+
+import (
+	"fmt"
+	"github.com/goombaio/namegenerator"
+	"github.com/hashicorp/go-uuid"
+	"time"
+)
+
+func GenTraceId() string {
```
Comment (jonalu) [inline src/buldog/trace_id.go]: Kan være greit å unitteste denne

File: src/buldog/cache.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+func (c Cache) Get(traceId string) []string {
+	entries, hasEntries := c.store[traceId]
+	if !hasEntries {
+		return []string{}
+	}
+	return entries
+}
+
+func (c *Cache) Clear(traceId string) {
+	c.store[traceId] = []string{}
```
[the remark this answers is not in the document]
Comment (eaardal) [reply src/buldog/cache.go]: kult, visste ikkje om den 👍

File: src/buldog/logger_context.go (hunk 1 above)
Comment (eaardal) [reply src/buldog/logger_context.go]: Desse tre sjekkane er rein paranoia :P Ingen av dei vil intreffe i praksis om ein bruker loggeren sånn den er tiltenkt. Vil heller at alt krasjer i slike ekstremt sære tilfeller enn å returnere ein gyldig verdi og aldri detektere at vi har et problem

File: src/buldog/logger_context_mock.go (hunk 2 above)
Comment (eaardal) [reply src/buldog/logger_context_mock.go]: Kan jo det men tipper Diagnostics(), Err(), og Value() aldri kjem til å bli brukt i praksis

File: src/buldog/logger_mock.go (hunk 3 above)
Comment (eaardal) [reply src/buldog/logger_mock.go]: den andre parameteren er _actual_ ja. Kan kanskje bruke named attrs om det blir synlig i autocompleten

**Pull request merged**